core/src/main/scala/org/apache/spark/util/collection/ExternalSorter.scala

```diff
@@ -740,15 +723,29 @@ private[spark] class ExternalSorter[K, V, C](
           in.close()
         }
       }
+    } else if (spills.isEmpty && partitionWriters == null) {
```

Comment:
The branching here is starting to get very complicated (#1799 added a second level of branching, and now this adds a third). Also, it a bit redundant with the branching in `partitionedIterator`, which also has its own special case for this. However, I don't seen an obvious way to improve it given the design of these other data structures. I'll keep thinking